src/coreclr/nativeaot/System.Private.CoreLib/src/System/Activator.CoreRT.cs

```diff
@@ -28,48 +28,55 @@ public static partial class Activator
-                return default;
-            }
-            else
+            // Grab the pointer to the default constructor of the type. If T doesn't have a default
+            // constructor, the intrinsic returns a marker pointer that we check for.
+            IntPtr defaultConstructor = DefaultConstructorOf<T>();
+
+            // Check if we got the marker back.
+            //
+            // TODO: might want to disambiguate the different cases for abstract class, interface, etc.
+            if (defaultConstructor == (IntPtr)(delegate*<Guid>)&MissingConstructorMethod)
+                throw new MissingMethodException(SR.Format(SR.MissingConstructor_Name, typeof(T)));
+
+            T t;
+            try
             {
-                // Grab the pointer to the default constructor of the type. If T doesn't have a default
-                // constructor, the intrinsic returns a marker pointer that we check for.
-                IntPtr defaultConstructor = DefaultConstructorOf<T>();
-
-                // Check if we got the marker back.
-                //
-                // TODO: might want to disambiguate the different cases for abstract class, interface, etc.
-                if (defaultConstructor == DefaultConstructorOf<ClassWithMissingConstructor>())
-                    throw new MissingMethodException(SR.Format(SR.MissingConstructor_Name, typeof(T)));
-
-                // Grab a pointer to the optimized allocator for the type and call it.
-                // TODO: we need RyuJIT to respect that RawCalliHelper doesn't do fat pointer transform
-                // IntPtr allocator = AllocatorOf<T>();
-                // T t = RawCalliHelper.Call<T>(allocator, EETypePtr.EETypePtrOf<T>().RawValue);
-                T t = (T)RuntimeImports.RhNewObject(EETypePtr.EETypePtrOf<T>());
-
-                try
+                // Call the default constructor on the allocated instance.
+                if (RuntimeHelpers.IsReference<T>())
                 {
-                    // Call the default constructor on the allocated instance.
+                    // Grab a pointer to the optimized allocator for the type and call it.
+                    IntPtr allocator = AllocatorOf<T>();
+                    t = RawCalliHelper.Call<T>(allocator, EETypePtr.EETypePtrOf<T>().RawValue);
                     RawCalliHelper.Call(defaultConstructor, t);
 
                     // Debugger goo so that stepping in works. Only affects debug info generation.
                     // The call gets optimized away.
                     DebugAnnotations.PreviousCallContainsDebuggerStepInCode();
-
-                    return t;
                 }
-                catch (Exception e)
+                else
                 {
-                    throw new TargetInvocationException(e);
+                    t = default;
+                    if (defaultConstructor != (IntPtr)(delegate*<Guid>)&ValueTypeWithNoConstructorMethod)
```

Comment:
We can have just an empty method that does nothing for this case to avoid the Guid hack. The check can then be just remove here.

Reply:
Are there no concerns about the calling convention of the method?

I wanted to do that, but I can't express a parameterless constructor in a struct in C# yet and right now there's a mismatch.

Reply:
Ok, you are right it would not be an obvious improvement.